cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java

```diff
@@ -111,14 +113,45 @@ protected void runOneIteration() throws Exception {
 
   @Override
   public void enforce(EntityId entity, Principal principal, Action action) throws Exception {
+    enforce(entity, principal, Collections.singleton(action));
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Set<Action> actions) throws Exception {
     if (!authorizationEnabled) {
       return;
     }
-    Set<Privilege> privileges = cacheEnabled ? authPolicyCache.get(principal) : getPrivileges(principal);
-    privileges = privileges == null ? ImmutableSet.<Privilege>of() : privileges;
-    if (!privileges.contains(new Privilege(entity, action))) {
-      throw new UnauthorizedException(principal, action, entity);
+    Set<Privilege> privileges = getPrivileges(principal);
+    // If a principal has ALL privileges on the entity, authorization should succeed
+    if (privileges.contains(new Privilege(entity, Action.ALL))) {
+      return;
+    }
+    // Otherwise check for the specific action requested
+    for (Action action : actions) {
+      if (!privileges.contains(new Privilege(entity, action))) {
+        throw new UnauthorizedException(principal, action, entity);
+      }
+    }
+  }
+
+  @Override
+  public <T extends EntityId> Set<T> filter(Set<T> unfiltered, Principal principal) throws Exception {
+    if (Principal.SYSTEM.equals(principal)) {
+      return unfiltered;
+    }
+
+    Set<Privilege> privileges = getPrivileges(principal);
```

Comment:
they are two different 'layers' in the code. ``InMemoryAuthorizer`` is the actual authorization backend ( in production it could be a ``SentryAuthorizer``). ``AuthorizationEnforcementService`` is a cache/wrapper around ``InMemoryAuthorizer``. I think it looks a little unclean right now because ``Authorizer`` also extends ``AuthorizationEnforcer``, which could be removed in future. When we do that, the ``InMemoryAuthorizer`` will no longer need a ``filter`` or ``enforce`` methods. But that change will require some more work in ``AuthorizerInstantiator``, which I'll defer to later.